File: src/polybot_control_plane/catalog/graphs/catalog.py
"""Public graph-node catalog descriptors and framework discovery assembly."""

from __future__ import annotations

from typing import Any, Literal, Self

from polybot_control_plane.catalog.graphs.ports import (
    GraphInputDescriptor,
    GraphOutputDescriptor,
)
from polybot_control_plane.catalog.graphs.preview_samples import (
    sample_payload,
    PREVIEW_SAMPLE_TIME_MS,
)
from polybot_control_plane.catalog.graphs.operations import (
    GraphOperationDescriptor,
    operation_descriptors,
)

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    StrictBool,
    StrictStr,
    model_validator,
)

from polybot.framework.base import BaseBot
from polybot.framework.events import Side
from polybot_control_plane.catalog.graphs._actions import (
    DiscoveredBrokerAction,
    discover_broker_actions,
)
from polybot_control_plane.catalog.graphs._fields import DiscoveredGraphField
from polybot_control_plane.catalog.graphs._triggers import (
    DiscoveredGraphPayload,
    DiscoveredGraphTrigger,
    discover_graph_triggers,
)
from polybot_control_plane.catalog.graphs._validation import (
    ensure_unique_graph_trigger_hooks,
    ensure_unique_values,
)
from polybot_control_plane.catalog.graphs.comparisons import GRAPH_COMPARISON_SPECS
from polybot_control_plane.catalog.graphs.values import (
    GRAPH_BROKER_SUBMIT_METHOD_NAME,
    GRAPH_COMPARISON_LEFT_HANDLE_ID,
    GRAPH_COMPARISON_RESULT_HANDLE_ID,
    GRAPH_COMPARISON_RIGHT_HANDLE_ID,
    GRAPH_CONTEXT_HANDLE_ID,
    GRAPH_CONTEXT_TYPE_NAME,
    GRAPH_VALUE_HANDLE_ID,
    GraphBrokerAction,
    GraphComparisonOperator,
    GraphNodeType,
    GraphScalarType,
    GraphPort,
)
from polybot_control_plane.catalog.graphs.types import (
    GraphFieldPath,
    GraphHookName,
    GraphHandleId,
)


class GraphFieldDescriptor(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    path: GraphFieldPath
    handle_id: GraphHandleId
    display_name: str
    value_type: str
    scalar_type: GraphScalarType | None
    nullable: bool
    collection: bool
    value_schema: dict[str, object]

    @classmethod
    def from_discovered(
        cls,
        payload_type_name: str,
        field: DiscoveredGraphField,
    ) -> Self:
        path = GraphFieldPath(segments=field.path)
        return cls(
            path=path,
            handle_id=path.handle_id,
            display_name=f"{payload_type_name}.{path.dotted}",
            value_type=field.value_type,
            scalar_type=field.scalar_type,
            nullable=field.nullable,
            collection=field.collection,
            value_schema=dict(field.value_schema),
        )


class GraphPayloadDescriptor(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    type_name: str
    fields: tuple[GraphFieldDescriptor, ...]

    @classmethod
    def from_discovered(cls, payload: DiscoveredGraphPayload) -> Self:
        return cls(
            type_name=payload.type_name,
            fields=tuple(
                GraphFieldDescriptor.from_discovered(payload.type_name, field)
                for field in payload.fields
            ),
        )

    def field_for_path(self, path: GraphFieldPath) -> GraphFieldDescriptor | None:
        return next((field for field in self.fields if field.path == path), None)


class GraphTriggerDescriptor(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    node_type: Literal[GraphNodeType.TRIGGER] = GraphNodeType.TRIGGER
    hook_name: GraphHookName
    context_handle_id: Literal[GRAPH_CONTEXT_HANDLE_ID]
    context_type_name: Literal[GRAPH_CONTEXT_TYPE_NAME]
    payload: GraphPayloadDescriptor | None = None
    sample_payload: dict[str, Any] | None = None
    sample_time_ms: int = PREVIEW_SAMPLE_TIME_MS

    @classmethod
    def from_discovered(cls, trigger: DiscoveredGraphTrigger) -> Self:
        return cls(
            hook_name=trigger.hook_name,
            sample_payload=sample_payload(trigger.hook_name),
            context_handle_id=GRAPH_CONTEXT_HANDLE_ID,
            context_type_name=GRAPH_CONTEXT_TYPE_NAME,
            payload=(
                None
                if trigger.payload is None
                else GraphPayloadDescriptor.from_discovered(trigger.payload)
            ),
        )


class GraphConstantDescriptor(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    node_type: Literal[GraphNodeType.CONSTANT] = GraphNodeType.CONSTANT
    scalar_type: GraphScalarType
    display_name: str
    default_value: StrictBool | StrictStr
    output: GraphOutputDescriptor

    @classmethod
    def from_scalar_type(
        cls,
        scalar_type: GraphScalarType,
        default_value: bool | str,
    ) -> Self:
        return cls(
            scalar_type=scalar_type,
            display_name=f"{'Text' if scalar_type is GraphScalarType.STRING else scalar_type.value.title()} constant",
            default_value=default_value,
            output=GraphOutputDescriptor(
                handle_id=GRAPH_VALUE_HANDLE_ID,
                display_name="Value",
                scalar_type=scalar_type,
            ),
        )


class GraphComparisonDescriptor(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    node_type: Literal[GraphNodeType.COMPARISON] = GraphNodeType.COMPARISON
    operator: GraphComparisonOperator
    display_name: str
    inputs: tuple[GraphInputDescriptor, GraphInputDescriptor]
    output: GraphOutputDescriptor

    @classmethod
    def from_operator(cls, operator: GraphComparisonOperator) -> Self:
        scalar_types = GRAPH_COMPARISON_SPECS[operator].scalar_types
        return cls(
            operator=operator,
            display_name=operator.value.replace("_", " ").title(),
            inputs=tuple(
                GraphInputDescriptor(
                    handle_id=handle_id,
                    display_name=handle_id.title(),
                    scalar_types=scalar_types,
                    nullable=True,
                    required=True,
                )
                for handle_id in (
                    GRAPH_COMPARISON_LEFT_HANDLE_ID,
                    GRAPH_COMPARISON_RIGHT_HANDLE_ID,
                )
            ),
            output=GraphOutputDescriptor(
                handle_id=GRAPH_COMPARISON_RESULT_HANDLE_ID,
                display_name="Result",
                scalar_type=GraphScalarType.BOOLEAN,
                nullable=True,
            ),
        )


class GraphBrokerActionDescriptor(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    node_type: Literal[GraphNodeType.BROKER_ACTION] = GraphNodeType.BROKER_ACTION
    action: GraphBrokerAction
    method_name: Literal[GRAPH_BROKER_SUBMIT_METHOD_NAME]
    display_name: str
    side: Side
    inputs: tuple[GraphInputDescriptor, ...]
    outputs: tuple[GraphOutputDescriptor, ...] = ()

    @classmethod
    def from_discovered(cls, action: DiscoveredBrokerAction) -> Self:
        return cls(
            action=GraphBrokerAction(
                f"{GRAPH_BROKER_SUBMIT_METHOD_NAME}_{action.side.value.lower()}"
            ),
            method_name=action.method_name,
            display_name=f"{action.side.value} order",
            side=action.side,
            outputs=tuple(
                GraphOutputDescriptor(
                    handle_id=name,
                    display_name="Execution Price" if name == GraphPort.AVERAGE_PRICE else name.replace("_", " ").title(),
                    scalar_type=kind,
                    nullable=name != GraphPort.STATUS,
                )
                for name, kind in (
                    (GraphPort.STATUS, GraphScalarType.STRING),
                    (GraphPort.FILLED_SIZE, GraphScalarType.NUMBER),
                    (GraphPort.AVERAGE_PRICE, GraphScalarType.NUMBER),
                    (GraphPort.SKIP_REASON, GraphScalarType.STRING),
                    (GraphPort.REJECT_REASON, GraphScalarType.STRING),
                )
            ),
            inputs=tuple(
                GraphInputDescriptor(
                    handle_id=input_.name,
                    display_name=input_.name.replace("_", " ").title(),
                    scalar_types=(input_.scalar_type,),
                    nullable=input_.nullable,
                    required=input_.required,
                )
                for input_ in action.inputs
            ),
        )


class GraphNodeCatalog(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    operations: tuple[GraphOperationDescriptor, ...] = ()
    triggers: tuple[GraphTriggerDescriptor, ...]
    constants: tuple[GraphConstantDescriptor, ...]
    comparisons: tuple[GraphComparisonDescriptor, ...]
    broker_actions: tuple[GraphBrokerActionDescriptor, ...]

    @model_validator(mode="after")
    def _validate_unique_entries(self) -> Self:
        ensure_unique_graph_trigger_hooks(
            tuple(trigger.hook_name for trigger in self.triggers),
        )
        ensure_unique_values(
            tuple(constant.scalar_type for constant in self.constants),
            "graph constant scalar type",
        )
        ensure_unique_values(
            tuple(comparison.operator for comparison in self.comparisons),
            "graph comparison operator",
        )
        ensure_unique_values(
            tuple(action.action for action in self.broker_actions),
            "graph broker action",
        )
        return self

    @classmethod
    def from_bot_type(cls, bot_type: type[Any]) -> Self:
        return cls(
            operations=operation_descriptors(),
            triggers=tuple(
                GraphTriggerDescriptor.from_discovered(trigger)
                for trigger in discover_graph_triggers(bot_type)
            ),
            constants=(
                GraphConstantDescriptor.from_scalar_type(
                    GraphScalarType.BOOLEAN,
                    False,
                ),
                GraphConstantDescriptor.from_scalar_type(
                    GraphScalarType.NUMBER,
                    "0",
                ),
                GraphConstantDescriptor.from_scalar_type(
                    GraphScalarType.STRING,
                    "",
                ),
            ),
            comparisons=tuple(
                GraphComparisonDescriptor.from_operator(operator)
                for operator in GraphComparisonOperator
            ),
            broker_actions=tuple(
                GraphBrokerActionDescriptor.from_discovered(action)
                for action in discover_broker_actions()
            ),
        )

    def trigger(self, hook_name: str) -> GraphTriggerDescriptor | None:
        return next(
            (trigger for trigger in self.triggers if trigger.hook_name == hook_name),
            None,
        )

    def constant(self, scalar_type: GraphScalarType) -> GraphConstantDescriptor:
        return next(item for item in self.constants if item.scalar_type is scalar_type)

    def comparison(
        self,
        operator: GraphComparisonOperator,
    ) -> GraphComparisonDescriptor:
        return next(item for item in self.comparisons if item.operator is operator)

    def broker_action(
        self,
        action: GraphBrokerAction,
    ) -> GraphBrokerActionDescriptor:
        return next(item for item in self.broker_actions if item.action is action)


GRAPH_NODE_CATALOG = GraphNodeCatalog.from_bot_type(BaseBot)

File: frontend/src/lib/api/responseValidation/graph/catalog.ts
import { isRecord, isArrayOf, isNonemptyString, isOneOf } from "$lib/valueGuards";
import { hasOnlyResponseFields } from "../fields";
import catalogContract from "$lib/catalog/catalogContract.fixture.json";
import runtimeContract from "$lib/runtimeContract.fixture.json";
import { graphScalarValueIsValid as isScalarValue } from "$lib/catalog/scalarValue";
import {
  GRAPH_SCALAR_TYPES,
  GRAPH_COMPARISON_OPERATORS,
  GRAPH_BROKER_ACTIONS,
  isGraphIdentifier,
  isGraphHookName,
  isGraphFieldPathSegment,
  hasUniqueValues,
} from "./primitives";

export function isGraphCatalog(value: unknown): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphNodeCatalog")) return false;
  if (
    !isArrayOf(value.triggers, isTriggerDescriptor) ||
    !isArrayOf(value.operations, isOperationDescriptor) ||
    !isArrayOf(value.constants, isConstantDescriptor) ||
    !isArrayOf(value.comparisons, isComparisonDescriptor) ||
    !isArrayOf(value.broker_actions, isBrokerActionDescriptor)
  )
    return false;
  return (
    hasUniqueValues(value.triggers, "hook_name") &&
    hasUniqueValues(value.operations, "operation") &&
    hasUniqueValues(value.constants, "scalar_type") &&
    hasUniqueValues(value.comparisons, "operator") &&
    hasUniqueValues(value.broker_actions, "action")
  );
}

function isTriggerDescriptor(value: Record<string, unknown>): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphTriggerDescriptor")) return false;
  const canonicalDescriptor = catalogContract.graphNodeCatalog.triggers.find(
    (descriptor) => descriptor.hook_name === value.hook_name,
  );
  return (
    canonicalDescriptor !== undefined &&
    value.node_type === canonicalDescriptor.node_type &&
    value.context_handle_id === canonicalDescriptor.context_handle_id &&
    value.context_type_name === canonicalDescriptor.context_type_name &&
    isGraphHookName(value.hook_name) &&
    (value.payload === undefined || value.payload === null || isGraphPayloadDescriptor(value.payload))
  );
}

function isGraphPayloadDescriptor(value: unknown): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphPayloadDescriptor")) return false;
  return isNonemptyString(value.type_name) && Array.isArray(value.fields) && value.fields.every(isGraphPayloadField);
}

function isConstantDescriptor(value: Record<string, unknown>): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphConstantDescriptor")) return false;
  const canonicalDescriptor = catalogContract.graphNodeCatalog.constants.find(
    (descriptor) => descriptor.scalar_type === value.scalar_type,
  );
  return (
    canonicalDescriptor !== undefined &&
    value.node_type === canonicalDescriptor.node_type &&
    isNonemptyString(value.display_name) &&
    isOneOf(value.scalar_type, GRAPH_SCALAR_TYPES) &&
    isScalarValue(value.scalar_type, value.default_value) &&
    isRecord(value.output) &&
    isGraphOutputDescriptor(value.output)
  );
}

function isComparisonDescriptor(value: Record<string, unknown>): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphComparisonDescriptor")) return false;
  const canonicalDescriptor = catalogContract.graphNodeCatalog.comparisons.find(
    (descriptor) => descriptor.operator === value.operator,
  );
  return (
    isNonemptyString(value.display_name) &&
    isOneOf(value.operator, GRAPH_COMPARISON_OPERATORS) &&
    canonicalDescriptor !== undefined &&
    value.node_type === canonicalDescriptor.node_type &&
    Array.isArray(value.inputs) &&
    value.inputs.length === canonicalDescriptor.inputs.length &&
    value.inputs.every(isGraphInputDescriptor) &&
    isRecord(value.output) &&
    isGraphOutputDescriptor(value.output)
  );
}

function isBrokerActionDescriptor(value: Record<string, unknown>): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphBrokerActionDescriptor")) return false;
  const canonicalDescriptor = catalogContract.graphNodeCatalog.broker_actions.find(
    (descriptor) => descriptor.action === value.action,
  );
  return (
    canonicalDescriptor !== undefined &&
    value.node_type === canonicalDescriptor.node_type &&
    isNonemptyString(value.display_name) &&
    isOneOf(value.action, GRAPH_BROKER_ACTIONS) &&
    value.method_name === catalogContract.graphBrokerSubmitMethodName &&
    isOneOf(value.side, Object.values(runtimeContract.side)) &&
    Array.isArray(value.inputs) &&
    value.inputs.every(isGraphInputDescriptor) &&
    (value.outputs === undefined || isArrayOf(value.outputs, isGraphOutputDescriptor))
  );
}

function isGraphInputDescriptor(value: unknown): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphInputDescriptor")) return false;
  return (
    isGraphIdentifier(value.handle_id) &&
    isNonemptyString(value.display_name) &&
    Array.isArray(value.scalar_types) &&
    value.scalar_types.length >= catalogContract.nodeGraph.minimumInputScalarTypes &&
    value.scalar_types.every((type) => isOneOf(type, [...GRAPH_SCALAR_TYPES, catalogContract.contextPortType])) &&
    typeof value.nullable === "boolean" &&
    typeof value.required === "boolean"
  );
}

function isGraphOutputDescriptor(value: Record<string, unknown>): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphOutputDescriptor")) return false;
  return (
    isGraphIdentifier(value.handle_id) &&
    isNonemptyString(value.display_name) &&
    isOneOf(value.scalar_type, [...GRAPH_SCALAR_TYPES, catalogContract.contextPortType]) &&
    (value.nullable === undefined || typeof value.nullable === "boolean")
  );
}

function isOperationDescriptor(value: Record<string, unknown>): boolean {
  if (!isRecord(value) || !hasOnlyResponseFields(value, "GraphOperationDescriptor")) return false;
  return (
    catalogContract.graphNodeCatalog.operations.some((item) => item.operation === value.operation) &&
    isNonemptyString(value.display_name) &&
    isNonemptyString(value.category) &&
    isArrayOf(value.inputs, isGraphInputDescriptor) &&
    isArrayOf(value.outputs, isGraphOutputDescriptor)
  );
}

function isGraphPayloadField(field: unknown): boolean {
  if (!isRecord(field) || !hasOnlyResponseFields(field, "GraphFieldDescriptor")) return false;
  return (
    isGraphIdentifier(field.handle_id) &&
    isNonemptyString(field.display_name) &&
    isRecord(field.path) &&
    hasOnlyResponseFields(field.path, "GraphFieldPath") &&
    Array.isArray(field.path.segments) &&
    field.path.segments.length >= catalogContract.nodeGraph.minimumFieldPathSegments &&
    field.path.segments.every(isGraphFieldPathSegment) &&
    isNonemptyString(field.value_type) &&
    (field.scalar_type === null || isOneOf(field.scalar_type, GRAPH_SCALAR_TYPES)) &&
    typeof field.nullable === "boolean" &&
    typeof field.collection === "boolean" &&
    isRecord(field.value_schema)
  );
}
